src/Lean/Elab/Tactic/BVDecide/Bitblast/BVExpr/Circuit/Lemmas/Add.lean

```diff
@@ -0,0 +1,245 @@
+    (lhsExpr rhsExpr : BitVec w)
+    (hleft : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, lhs.get idx hidx, assign⟧ = lhsExpr.getLsb idx)
+    (hright : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, rhs.get idx hidx, assign⟧ = rhsExpr.getLsb idx)
+    (hcin : ⟦aig, cin, assign⟧ = BitVec.carry curr lhsExpr rhsExpr false) :
+    ∀ (idx : Nat) (hidx1 : idx < w),
+        curr ≤ idx
+          →
+        ⟦
+          (go aig lhs rhs curr hcurr cin s).aig,
+          (go aig lhs rhs curr hcurr cin s).vec.get idx hidx1,
+          assign
+        ⟧
+          =
+        ⟦aig, lhs.get idx hidx1, assign⟧.xor
+          (⟦aig, rhs.get idx hidx1, assign⟧.xor
+            (BitVec.carry idx lhsExpr rhsExpr false)) := by
+  intro idx hidx1 hidx2
+  generalize hgo : go aig lhs rhs curr hcurr cin s = res
+  unfold go at hgo
+  dsimp only at hgo
+  split at hgo
+  . next hlt =>
+    cases Nat.eq_or_lt_of_le hidx2 with
+    | inl heq =>
+      rw [← hgo]
+      rw [go_get (hidx := by omega)]
+      rw [AIG.RefVec.get_push_ref_eq' (hidx := by rw [heq])]
+      simp only [← heq]
+      rw [go_denote_mem_prefix]
+      . unfold mkFullAdder
+        simp [hcin]
+      . simp only [Ref.gate_cast]
+        apply Ref.hgate
+    | inr hlt =>
+      rw [← hgo]
+      rw [go_denote_eq (lhsExpr := lhsExpr) (rhsExpr := rhsExpr) (curr := curr + 1)]
+      . rw [mkFullAdder_denote_mem_prefix]
+        rw [mkFullAdder_denote_mem_prefix]
+        . simp
+        . simp [Ref.hgate]
+        . simp [Ref.hgate]
+      . intro idx hidx
+        rw [mkFullAdder_denote_mem_prefix]
+        rw [← hleft idx hidx]
+        . simp
+        . simp [Ref.hgate]
+      . intro idx hidx
+        rw [mkFullAdder_denote_mem_prefix]
+        rw [← hright idx hidx]
+        . simp
+        . simp [Ref.hgate]
+      . unfold mkFullAdder
+        simp only [Ref.cast_eq, id_eq, Int.reduceNeg, denote_projected_entry, denote_mkFullAdderCarry,
+          FullAdderInput.lhs_cast, FullAdderInput.rhs_cast, FullAdderInput.cin_cast,
+          BitVec.carry_succ]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [hleft, hright, hcin]
+        simp [atLeastTwo_eq_halfAdder]
+      . omega
+  . omega
+termination_by w - curr
+
+end blastAdd
+
+theorem blastAdd_denote_eq (aig : AIG α) (lhs rhs : BitVec w) (assign : α → Bool)
+      (input : BinaryRefVec aig w)
+      (hleft : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, input.lhs.get idx hidx, assign⟧ = lhs.getLsb idx)
+      (hright : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, input.rhs.get idx hidx, assign⟧ = rhs.getLsb idx) :
+      ∀ (idx : Nat) (hidx : idx < w),
+          ⟦(blastAdd aig input).aig, (blastAdd aig input).vec.get idx hidx, assign⟧
+            =
+          (lhs + rhs).getLsb idx := by
+  intro idx hidx
+  rw [BitVec.getLsb_add]
+  . rw [← hleft idx hidx]
```

Comment:
Took me a while to notice you're using `.` here instead of `·`. I'm pretty sure that the dominant style in `Init` and `Lean` (and definitely in `Std`) is to use the center dot.